PWGCF/MultiparticleCorrelations/Tasks/multiparticle-correlations-ar.cxx

```diff
@@ -385,33 +610,505 @@ struct MultiParticleCorrelationsARTask {
+      2. * Q(n3, 1) * Q(n1 + n4, 2) * Q(n2 + n5 + n6, 3) -
+      2. * Q(n1, 1) * Q(n3 + n4, 2) * Q(n2 + n5 + n6, 3) +
+      4. * Q(n1 + n3 + n4, 3) * Q(n2 + n5 + n6, 3) -
+      6. * Q(n3, 1) * Q(n4, 1) * Q(n1 + n2 + n5 + n6, 4) +
+      6. * Q(n3 + n4, 2) * Q(n1 + n2 + n5 + n6, 4) +
+      2. * Q(n1, 1) * Q(n2, 1) * Q(n4, 1) * Q(n3 + n5 + n6, 3) -
+      2. * Q(n1 + n2, 2) * Q(n4, 1) * Q(n3 + n5 + n6, 3) -
+      2. * Q(n2, 1) * Q(n1 + n4, 2) * Q(n3 + n5 + n6, 3) -
+      2. * Q(n1, 1) * Q(n2 + n4, 2) * Q(n3 + n5 + n6, 3) +
+      4. * Q(n1 + n2 + n4, 3) * Q(n3 + n5 + n6, 3) -
+      6. * Q(n2, 1) * Q(n4, 1) * Q(n1 + n3 + n5 + n6, 4) +
+      6. * Q(n2 + n4, 2) * Q(n1 + n3 + n5 + n6, 4) -
+      6. * Q(n1, 1) * Q(n4, 1) * Q(n2 + n3 + n5 + n6, 4) +
+      6. * Q(n1 + n4, 2) * Q(n2 + n3 + n5 + n6, 4) +
+      24. * Q(n4, 1) * Q(n1 + n2 + n3 + n5 + n6, 5) +
+      2. * Q(n1, 1) * Q(n2, 1) * Q(n3, 1) * Q(n4 + n5 + n6, 3) -
+      2. * Q(n1 + n2, 2) * Q(n3, 1) * Q(n4 + n5 + n6, 3) -
+      2. * Q(n2, 1) * Q(n1 + n3, 2) * Q(n4 + n5 + n6, 3) -
+      2. * Q(n1, 1) * Q(n2 + n3, 2) * Q(n4 + n5 + n6, 3) +
+      4. * Q(n1 + n2 + n3, 3) * Q(n4 + n5 + n6, 3) -
+      6. * Q(n2, 1) * Q(n3, 1) * Q(n1 + n4 + n5 + n6, 4) +
+      6. * Q(n2 + n3, 2) * Q(n1 + n4 + n5 + n6, 4) -
+      6. * Q(n1, 1) * Q(n3, 1) * Q(n2 + n4 + n5 + n6, 4) +
+      6. * Q(n1 + n3, 2) * Q(n2 + n4 + n5 + n6, 4) +
+      24. * Q(n3, 1) * Q(n1 + n2 + n4 + n5 + n6, 5) -
+      6. * Q(n1, 1) * Q(n2, 1) * Q(n3 + n4 + n5 + n6, 4) +
+      6. * Q(n1 + n2, 2) * Q(n3 + n4 + n5 + n6, 4) +
+      24. * Q(n2, 1) * Q(n1 + n3 + n4 + n5 + n6, 5) +
+      24. * Q(n1, 1) * Q(n2 + n3 + n4 + n5 + n6, 5) -
+      120. * Q(n1 + n2 + n3 + n4 + n5 + n6, 6);
+    return six;
+  }
+
+  TComplex Recursion(int n, int* harmonic, int mult = 1, int skip = 0)
+  {
+    // Calculate multi-particle correlators by using recursion (an improved
+    // faster version) originally developed by Kristjan Gulbrandsen
+    // (gulbrand@nbi.dk).
+    int nm1 = n - 1;
+    TComplex c(Q(harmonic[nm1], mult));
+    if (nm1 == 0)
+      return c;
+    c *= Recursion(nm1, harmonic);
+    if (nm1 == skip)
+      return c;
+    int multp1 = mult + 1;
+    int nm2 = n - 2;
+    int counter1 = 0;
+    int hhold = harmonic[counter1];
+    harmonic[counter1] = harmonic[nm2];
+    harmonic[nm2] = hhold + harmonic[nm1];
+    TComplex c2(Recursion(nm1, harmonic, multp1, nm2));
+    int counter2 = n - 3;
+    while (counter2 >= skip) {
+      harmonic[nm2] = harmonic[counter1];
+      harmonic[counter1] = hhold;
+      ++counter1;
+      hhold = harmonic[counter1];
+      harmonic[counter1] = harmonic[nm2];
+      harmonic[nm2] = hhold + harmonic[nm1];
+      c2 += Recursion(nm1, harmonic, multp1, counter2);
+      --counter2;
+    }
+    harmonic[nm2] = harmonic[counter1];
+    harmonic[counter1] = hhold;
+    if (mult == 1)
+      return c - c2;
+    return c - double(mult) * c2;
+  }
+
```

Comment:
Couldn't you use the Generic Framework already implemented?

Reply:
To be honest, I have not really looked at their code. Ante and myself we would like to have a code seperate from theirs.

Reply:
I'm not quite sure about this reason
O2Physics is not about personal opinions but about an effective Collaboration framework